src/ray/core_worker/task_event_buffer.h

```diff
@@ -66,10 +70,99 @@ class TaskEventBuffer {
   ///       CoreWorker disconnects to ensure all task events in the buffer are sent.
   virtual void FlushEvents(bool forced) = 0;
 
+  /// Start the TaskEventBuffer.
+  ///
+  /// Connects the GCS client, starts its io_thread, and sets up periodical runner for
+  /// flushing events to GCS.
+  /// \param auto_flush Test only flag to disable periodical flushing events if false.
+  /// \return If setup succeeds. When failure, events will not be reported.
+  virtual Status Start(bool auto_flush = true) = 0;
+
   /// Stop the TaskEventBuffer and it's underlying IO, disconnecting GCS clients.
   virtual void Stop() = 0;
 };
 
+/// Implementation of TaskEventBuffer.
+///
+/// The buffer has its own io_context and io_thread, that's isolated from other
+/// components.
+///
+/// This class is thread-safe.
+class TaskEventBufferImpl : public TaskEventBuffer {
+ public:
+  /// Constructor
+  ///
+  /// \param gcs_client GCS client
+  TaskEventBufferImpl(std::unique_ptr<gcs::GcsClient> gcs_client);
+
+  void AddTaskEvents(rpc::TaskEvents task_events) LOCKS_EXCLUDED(mutex_) override;
```

Comment:
```suggestion
  void AddTaskEvent(rpc::TaskEvent task_event) LOCKS_EXCLUDED(mutex_) override;
```

Isn't it a single event?

Reply:
Ah yeah - this is something i feel a bit weird as well. Technically we are adding a single event each time - but the protobuf itself might contain multiple events in a `rpc::TaskEvent` (e.g. multiple profile events).

Reply:
So there will also be multiple events in a `rpc::TaskEvents` (initially we will have multiple `status events`). 

But maybe we could also flatten them out? I guess this will have some fields being duplicated (metadata that's shared by each profile event).